### PR TITLE
Cmake rework

### PR DESCRIPTION
**Pull request by JensWehner**
- Pushes Cmake to 3.10

Todo
- [x] Refactor pkg config into packages

Reworked the target definitions for 
 - [x] BLAS
  - [x] LAPACK
 - [x] MPI
 - [x] QHULL (https://github.com/hpgem/hpgem/issues/25)
 - [x] METIS
 - [x] PETSC
 - [x] SLEPC
- [x] Sundial (https://github.com/hpgem/hpgem/issues/24)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required (VERSION 3.3)
+cmake_minimum_required (VERSION 3.10)
 project (hpGEM C CXX Fortran)
```
Comment (Laar) [inline CMakeLists.txt:2]: I think we are not using Fortran anymore now that our inline blas and lapack have been removed.

```suggestion
project (hpGEM C CXX)
```

**Comment by JensWehner** [comment]
@Laar Can you try this out?

File: conf/cmake/ConfigureDependencies.cmake [hunk 2]
```diff
@@ -3,93 +3,78 @@
...
 
+
+
 ### OPTIONAL DEPENDENCIES ###
 #############################
 
 if(hpGEM_USE_MPI)
-    FIND_PACKAGE(MPI REQUIRED)
-    # Add OMPI_SKIP_MPICXX to exclude the deprecated c++ bindings of MPI
-    add_definitions(-DHPGEM_USE_MPI -DOMPI_SKIP_MPICXX)
```
Comment (Laar) [inline conf/cmake/ConfigureDependencies.cmake:34]: The `-DOMPI_SKIP_MPICXX` seems to be lost. Is this intentional?

**Comment by Laar** [comment]
My trouble so far with the new pkg-config approach

== OSX ==
 - pkg-config is not installed by default
 - After installing it does not find the user installed petsc with `PETSC_DIR=~/[path]` and `PETSC_ARCH=[arch]` set as environment variables.

== Arch linux ==
 - Using real petsc from AUR, with PETSC_DIR, SLEPC_DIR as environment variables and PETSC_ARCH unset, it finds PETSc but does not find SLEPc
 - Using a modified version of the same AUR for complex numbers and similar environment variables it picks up the real PETSc and fails.

So in both cases I do not know if there is a way to point it to the correct location, but so far it does not work.

**Comment by JensWehner** [comment]
@Laar I updated the cmake now you can specify SLEPc_LIBRARY and SLEPc_INCLUDE_DIR and the same for PETSc

File: conf/cmake/ConfigureDependencies.cmake [hunk 3]
```diff
@@ -3,167 +3,111 @@
...
-        message(FATAL_ERROR "The option you have chosen requires QHull and you do not have this installed. Please install")
-    endif()
     # Create target for easy linking
     add_library(QHULL::QHULL INTERFACE IMPORTED)
-    target_include_directories(QHULL::QHULL INTERFACE "${QHULL_INCLUDE_DIR}")
     if(CMAKE_BUILD_TYPE MATCHES "Debug")
-        target_link_libraries(QHULL::QHULL INTERFACE "${QHULL_DEBUG_LIBRARIES}")
+        set_target_properties(QHULL::QHULL PROPERTIES 
+        INTERFACE_LINK_LIBRARIES "${QHULL_DEBUG_LIBRARIES}" 
+        INTERFACE_INCLUDE_DIRECTORIES "${QHULL_INCLUDE_DIR}")
```
Comment (Laar) [inline conf/cmake/ConfigureDependencies.cmake:48]:
```suggestion
        set_target_properties(QHULL::QHULL PROPERTIES 
            INTERFACE_LINK_LIBRARIES "${QHULL_DEBUG_LIBRARIES}" 
            INTERFACE_INCLUDE_DIRECTORIES "${QHULL_INCLUDE_DIR}")
```
Indentation suggestion

File: conf/cmake/ConfigureDependencies.cmake [hunk 4]
```diff
@@ -3,167 +3,111 @@
...
     else()
-        target_link_libraries(QHULL::QHULL INTERFACE "${QHULL_LIBRARIES}")
+        set_target_properties(QHULL::QHULL PROPERTIES 
+        INTERFACE_LINK_LIBRARIES "${QHULL_LIBRARIES}" 
+        INTERFACE_INCLUDE_DIRECTORIES "${QHULL_INCLUDE_DIR}")
```
Comment (Laar) [inline conf/cmake/ConfigureDependencies.cmake:52]:
```suggestion
        set_target_properties(QHULL::QHULL PROPERTIES 
            INTERFACE_LINK_LIBRARIES "${QHULL_LIBRARIES}" 
            INTERFACE_INCLUDE_DIRECTORIES "${QHULL_INCLUDE_DIR}")
```
Indentation suggestion

File: kernel/CMakeLists.txt [hunk 5]
```diff
@@ -2,7 +2,7 @@ add_library(Logger SHARED Logger.cpp)
 set_target_properties(Logger PROPERTIES POSITION_INDEPENDENT_CODE true)
 
 if( hpGEM_USE_MPI )
-  target_link_libraries(Logger "${MPI_LIBRARIES}")
+  target_link_libraries(Logger PRIVATE MPI::MPI_CXX)
```
Comment (Laar) [inline kernel/CMakeLists.txt:5]: Might this also be necessary for the other kernel targets? (Might have forgotten this in the dependency graph)

File: conf/cmake/Modules/FindSLEPc.cmake [hunk 6]
```diff
@@ -1,278 +1,50 @@
...
+  include(CheckSymbolExists)
+  foreach(SLEPc_dir ${SLEPc_INCLUDE_DIR})
+      
+      set(SLEPc_SLEPcconf_path "${SLEPc_dir}/slepcconf.h")
+      if(EXISTS "${SLEPc_SLEPcconf_path}")
+        file(STRINGS ${SLEPc_SLEPcconf_path} _contents REGEX "#define SLEPC_PETSC_ARCH[ \\t]+")
+      if(_contents)
+          string(REGEX REPLACE ".*#define SLEPC_PETSC_ARCH[ \\t]+\"?([^\\t \"]+)\"?" "\\1" SLEPc_ARCH "${_contents}")
+          message(STATUS "${SLEPc_ARCH}")
+      endif()
```
Comment (Laar) [inline conf/cmake/Modules/FindSLEPc.cmake:32]: Indentation on this block made it hard to read, might be improved.

File: conf/cmake/ConfigureDependencies.cmake [hunk 7]
```diff
@@ -3,167 +3,111 @@
...
-                "The option you have chosen requires SLEPc and you do not have this installed. Please install")
-    endif()
-    # Create target for easy linking
-    if(SLEPC_FOUND)
-        add_library(SLEPc::SLEPc INTERFACE IMPORTED)
-        target_include_directories(SLEPc::SLEPc INTERFACE "${SLEPC_INCLUDE_DIRS}")
-        target_link_libraries(SLEPc::SLEPc INTERFACE "${SLEPC_LIBRARIES}")
+   
+    if(NOT(SLEPc_ARCH STREQUAL PETSc_ARCH))
+        message(FATAL_ERROR "SLEPc and PETSc architecture differ. ${SLEPc_ARCH} vs ${PETSc_ARCH}")
```
Comment (Laar) [inline conf/cmake/ConfigureDependencies.cmake:95]: I get an error here:
```
CMake Error at conf/cmake/ConfigureDependencies.cmake:95 (message):
  SLEPc and PETSc architecture differ.  #define SLEPC_PETSC_ARCH "" vs
  #define PETSC_ARCH ""
```
Might it be that the parsing is done incorrectly and that `PETSC_ARCH` variant contains a newline at the start.

File: conf/cmake/ConfigureDependencies.cmake [hunk 8]
```diff
@@ -3,167 +3,111 @@
 #############################
 
 FIND_PACKAGE(BLAS REQUIRED)
-if(NOT BLAS_FOUND)
-    message(SYSTEM "Could not find external BLAS, using reference implementation instead")
-    set(hpGEM_USE_EXTERNAL_BLAS OFF CACHE BOOL "Try to use a version of BLAS optimised for your system" FORCE)
-else()
-    # Linking target
+if(BLAS_FOUND AND NOT TARGET BLAS::BLAS)
```
[the remark this answers is not in the document]
Comment (JensWehner) [reply conf/cmake/ConfigureDependencies.cmake:6]: true

File: conf/cmake/ConfigureDependencies.cmake (hunk 7 above)
Comment (JensWehner) [reply conf/cmake/ConfigureDependencies.cmake:95]: yeah I think the regex fails on empty ARCH

File: kernel/CMakeLists.txt (hunk 5 above)
Comment (JensWehner) [reply kernel/CMakeLists.txt:5]: I think the Logger does not care about the rest, but I do not know.

File: CMakeLists.txt [hunk 9]
```diff
@@ -19,14 +20,14 @@ endif(NOT CMAKE_BUILD_TYPE)
 
 
 #allow dynamically locating dynamic libraries if cmake and the OS support it
-if(${CMAKE_VERSION} VERSION_GREATER 2.8.12)
-    set(CMAKE_MACOSX_RPATH 1)
-endif()
-
-enable_testing()
-
+set(CMAKE_MACOSX_RPATH 1)
```
[the remark this answers is not in the document]
Comment (JensWehner) [reply CMakeLists.txt:23]: agreed

**Pull request merged**